File: app/src/main/java/com/intacta/sosviagens/Utils/Alerts.java
package com.intacta.sosviagens.Utils;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.provider.Settings;
import android.view.View;
import android.view.Window;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.firebase.ui.auth.AuthUI;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.intacta.sosviagens.Adapter.RecyclerMoreThenOneAdapter;
import com.intacta.sosviagens.Beans.Comment;
import com.intacta.sosviagens.Beans.Road;
import com.intacta.sosviagens.Beans.Suggest;
import com.intacta.sosviagens.Database.DBUtilities;
import com.intacta.sosviagens.R;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Objects;

import hyogeun.github.com.colorratingbarlib.ColorRatingBar;

import static com.intacta.sosviagens.Utils.Utilities.RC_SIGN_IN;

public class Alerts {
    private Activity activity;

    private int theme;
    private TextView title;
    private ProgressBar loading;
    private LinearLayout ratinglayout;
    private ColorRatingBar rating;
    private TextInputLayout commentlayout;
    private EditText comment;
    private TextInputLayout cortesylayout;
    private EditText cortesy;
    private EditText time;
    private Button save;
    private LinearLayout timelayout;
    private Spinner timetype;


    public Alerts(Activity activity, int theme) {
        this.activity = activity;
        this.theme = theme;
    }

    private boolean checkuser() {
        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        return user != null;
    }


    public void noConcess(final String address) {
        Query reference = FirebaseDatabase.getInstance().getReference("suggestion")
                .orderByChild("rodovia").equalTo(address);

        reference.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if(!dataSnapshot.exists() && !address.contains("Rua") || !address.contains("rua") ){
                    final AlertDialog.Builder builder = new AlertDialog.Builder(activity);
                    builder.setMessage("Esse endereço " + address + " não está no nosso sistema, caso você acredite que deveria estar, aperte Ok")
                            .setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    DBUtilities dbUtilities = new DBUtilities(activity);
                                    Suggest suggest = new Suggest();
                                    suggest.setRodovia(address);
                                    suggest.setDia(getDia());
                                    dbUtilities.SendSuggest(suggest);
                                }
                            })
                            .setNegativeButton("Cancelar", null);
                    AlertDialog dialog = builder.create();
                    dialog.show();
                    dialog.getButton(AlertDialog.BUTTON_NEGATIVE).setTextColor(activity.getResources().getColor(R.color.black));
                    dialog.getButton(AlertDialog.BUTTON_POSITIVE).setTextColor(activity.getResources().getColor(R.color.colorAccent));
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });





    }


    public void buildAlertMessageNoGps() {
        final AlertDialog.Builder builder = new AlertDialog.Builder(activity, theme);
        builder.setMessage("Seu GPS está desabilitado, ative-o para que possamos ajudar")
                .setCancelable(false)
                .setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                    public void onClick(@SuppressWarnings("unused") final DialogInterface dialog, @SuppressWarnings("unused") final int id) {
                        activity.startActivity(new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS));
                    }
                })
                .setNegativeButton("Não", new DialogInterface.OnClickListener() {
                    public void onClick(final DialogInterface dialog, @SuppressWarnings("unused") final int id) {
                        dialog.cancel();
                    }
                });
        final AlertDialog alert = builder.create();
        alert.show();
    }

    public void CommentAlert(final Road road) {
        final Animation out = AnimationUtils.loadAnimation(activity, R.anim.fui_slide_out_left);
        final Animation in = AnimationUtils.loadAnimation(activity, R.anim.fui_slide_in_right);
        if (checkuser()) {
            final BottomSheetDialog myDialog = new BottomSheetDialog(Objects.requireNonNull(activity), R.style.Dialog_No_Border);
            myDialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
            myDialog.setContentView(R.layout.comment_dialog);
            initView(myDialog);
            if (!myDialog.isShowing()) {
                myDialog.show();
            }
            save.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Salvar(comment,cortesy,time,rating,road,myDialog);
                }
            });

        } else {
            List<AuthUI.IdpConfig> providers = Collections.singletonList(
                    new AuthUI.IdpConfig.GoogleBuilder().build());
            activity.startActivityForResult(AuthUI.getInstance().createSignInIntentBuilder()
                    .setLogo(R.mipmap.ic_launcher)
                    .setAvailableProviders(providers)
                    .setTheme(R.style.AppTheme)
                    .build(), RC_SIGN_IN);
        }


    }

    private void Salvar(EditText comment, EditText cortesy, EditText time, ColorRatingBar rating, Road road, BottomSheetDialog dialog) {
        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        String username = user.getDisplayName();
        String comnt = comment.getText().toString();
        String oppinion = String.valueOf(rating.getRating());
        String tempo = time.getText().toString() + timetype.getSelectedItem().toString();
        String dia = getDia();
        Comment cb = new Comment(comnt, username, dia, cortesy.getText().toString(),tempo, oppinion,road.getId());
        cb.setRodovia(road.getRodovia());
        cb.setConcessionaria(road.getConcessionaria());
        DBUtilities commentsDB = new DBUtilities(activity);
        commentsDB.SendComment(cb);
        dialog.dismiss();
    }


    public void MoreConcess(final ArrayList<Road> roads) {
        final ArrayList<Road> options = new ArrayList<>(roads);
        final BottomSheetDialog myDialog = new BottomSheetDialog(Objects.requireNonNull(activity), R.style.Dialog_No_Border);
        myDialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        myDialog.setContentView(R.layout.morethenonelayout);
        if (myDialog.isShowing()) {
            System.out.println("Alerta sendo exibido");
            return;

        }
        RecyclerView roadssuggestion = myDialog.findViewById(R.id.roads);

        Button cancelar = myDialog.findViewById(R.id.cancel);
        cancelar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                myDialog.dismiss();
            }
        });
        if (roadssuggestion.getChildCount() > 0) {
            myDialog.show();
            return;
        }
        RecyclerMoreThenOneAdapter recyclerMoreThenOneAdapter = new RecyclerMoreThenOneAdapter(activity, options);
        GridLayoutManager llm = new GridLayoutManager(activity, 1, RecyclerView.VERTICAL, false);

        roadssuggestion.setAdapter(recyclerMoreThenOneAdapter);
        roadssuggestion.setLayoutManager(llm);

        myDialog.setOnDismissListener(new DialogInterface.OnDismissListener() {
            @Override
            public void onDismiss(DialogInterface dialog) {
                options.clear();
            }
        });
        myDialog.show();

    }


    private String getDia() {
        Date datenow = Calendar.getInstance().getTime();
        @SuppressLint("SimpleDateFormat") DateFormat df = new SimpleDateFormat("dd/MM/yyyy");
        return df.format(datenow);
    }


    private void initView(BottomSheetDialog myDialog) {
        title = myDialog.findViewById(R.id.title);
        loading = myDialog.findViewById(R.id.loading);
        ratinglayout = myDialog.findViewById(R.id.ratinglayout);
        rating = myDialog.findViewById(R.id.rating);
        commentlayout = myDialog.findViewById(R.id.commentlayout);
        comment = myDialog.findViewById(R.id.comment);
        cortesylayout = myDialog.findViewById(R.id.cortesylayout);
        cortesy = myDialog.findViewById(R.id.cortesy);
        time = myDialog.findViewById(R.id.time);
        save = myDialog.findViewById(R.id.save);
        timelayout = myDialog.findViewById(R.id.timelayout);
        timetype = myDialog.findViewById(R.id.timetype);
    }
}

File: app/src/main/java/com/intacta/sosviagens/Utils/Alerts.kt
package com.intacta.sosviagens.Utils

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Dialog
import android.content.Intent
import android.provider.Settings
import android.view.Window
import android.view.animation.AnimationUtils
import android.widget.*
import androidx.appcompat.app.AlertDialog

import com.firebase.ui.auth.AuthUI
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.intacta.sosviagens.model.Beans.Comment
import com.intacta.sosviagens.model.Beans.Road
import com.intacta.sosviagens.model.Beans.Suggest
import com.intacta.sosviagens.model.Database.DBUtilities
import com.intacta.sosviagens.R

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Objects

import hyogeun.github.com.colorratingbarlib.ColorRatingBar

import com.intacta.sosviagens.Utils.Utilities.RC_SIGN_IN
import de.mateware.snacky.Snacky

class Alerts(private val activity: Activity){



    private val dia: String
        get() {
            val datenow = Calendar.getInstance().time
            @SuppressLint("SimpleDateFormat") val df = SimpleDateFormat("dd/MM/yyyy")
            return df.format(datenow)
        }

    private fun checkuser(): Boolean {
        val user = FirebaseAuth.getInstance().currentUser
        return user != null
    }


    fun noConcess(address: String) {
        val reference = FirebaseDatabase.getInstance().getReference("suggestion")
                .orderByChild("rodovia").equalTo(address)

        reference.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (!dataSnapshot.exists() && !address.contains("Rua",true) || !address.contains("avenida",true)) {
                    val builder = AlertDialog.Builder(activity)
                    builder.setMessage("Esse endereço $address não está no nosso sistema, caso você acredite que deveria estar, aperte Ok")
                            .setPositiveButton("Ok") { dialog, which ->
                                val dbUtilities = DBUtilities(activity)
                                val suggest = Suggest(dia,address,"android")
                                dbUtilities.SendSuggest(suggest)
                            }
                            .setNegativeButton("Cancelar", null)
                    val dialog = builder.create()
                    dialog.show()
                    dialog.getButton(AlertDialog.BUTTON_NEGATIVE).setTextColor(activity.resources.getColor(R.color.black))
                    dialog.getButton(AlertDialog.BUTTON_POSITIVE).setTextColor(activity.resources.getColor(R.color.colorAccent))
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })


    }


    fun buildAlertMessageNoGps() {
        /*val dialog = Dialog(activity,R.style.Dialog_No_Border)
        //val layoutBinder = DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(dialog.getContext()),R.layout.alert_layout,null,false);
        dialog.setContentView(R.layout.alert_layout)
        dialog.setCanceledOnTouchOutside(true)
        val title = dialog.findViewById<TextView>(R.id.atitle)
        val message = dialog.findViewById<TextView>(R.id.message)
        val button = dialog.findViewById<Button>(R.id.defaultbutton)

        title.text = "GPS desativado"
        message.text = "Ative seu GPS para que possamos ajudar a encontrar números de emrgência"
        button.setOnClickListener {
            dialog.dismiss()
        }
        dialog.show()
        */

        Snacky.builder().setActivity(activity).error().setText("Erro ao obter localização GPS desativado")
                .setAction("OK") {
                    activity.startActivity( Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS))
                }.setDuration(BaseTransientBottomBar.LENGTH_INDEFINITE).show()


    }

    fun CommentAlert(id: String) {
        if (checkuser()) {
            val myDialog = BottomSheetDialog(Objects.requireNonNull(activity))
            myDialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
            myDialog.setCanceledOnTouchOutside(true)
            myDialog.setContentView(R.layout.comment_dialog)
            if (!myDialog.isShowing) {
                myDialog.show()
            }
            val title:TextView = myDialog.findViewById(R.id.title)!!
            val rating:ColorRatingBar = myDialog.findViewById(R.id.rating)!!
            val comment:EditText = myDialog.findViewById(R.id.comment)!!
            val save:Button = myDialog.findViewById(R.id.save)!!

            title.text = "Avalie seu atendimento"

            save.setOnClickListener { Salvar(comment,rating, id, myDialog) }

        } else {
            val providers = listOf(AuthUI.IdpConfig.GoogleBuilder().build())
            activity.startActivityForResult(AuthUI.getInstance().createSignInIntentBuilder()
                    .setLogo(R.mipmap.ic_launcher)
                    .setAvailableProviders(providers)
                    .setTheme(R.style.AppTheme)
                    .build(), RC_SIGN_IN)
        }


    }

    private fun Salvar(comment: EditText,rating: ColorRatingBar, id: String, dialog: BottomSheetDialog) {


        val user = FirebaseAuth.getInstance().currentUser
        val username = user!!.displayName
        val comnt = comment.text.toString()
        val oppinion = rating.rating.toString()
        //val tempo = time.text.toString() + timetype!!.selectedItem.toString()
        val dia = dia
        val cb = Comment(comnt,username!!,dia,oppinion,id)

        val commentsDB = DBUtilities(activity)
        commentsDB.SendComment(cb)
        dialog.dismiss()
    }


}



